modules/renter/download.go

```diff
@@ -519,3 +521,74 @@ func (r *Renter) DownloadHistory() []modules.DownloadInfo {
 	}
 	return downloads
 }
+
+// ClearDownloadHistory clears the renter's download history inclusive
+// of the provided before and after timestamps
+func (r *Renter) ClearDownloadHistory(before, after time.Time) error {
+	if err := r.tg.Add(); err != nil {
+		return err
+	}
+	defer r.tg.Done()
+	r.downloadHistoryMu.Lock()
+	defer r.downloadHistoryMu.Unlock()
+
+	// Timestamp validation
+	if before.Before(after) && !before.IsZero() {
+		return errors.New("before timestamp can not be newer then after timestamp")
+	}
+
+	// Clear download history if both before and after timestamps are zero values
+	if before.IsZero() && after.IsZero() {
+		r.downloadHistory = r.downloadHistory[:0]
+		return nil
+	}
+
+	// If timestamp is zero, we manually set the index, otherwise we search for the timestamp
+	var beforeIndex, afterIndex int
+	if before.IsZero() {
+		beforeIndex = len(r.downloadHistory) - 1
+	} else {
+		beforeIndex = sort.Search(len(r.downloadHistory), func(i int) bool {
+			return r.downloadHistory[i].staticStartTime.UnixNano() >= before.UnixNano()
+		})
+	}
+	if after.IsZero() {
+		afterIndex = 0
+	} else {
+		afterIndex = sort.Search(len(r.downloadHistory), func(i int) bool {
+			return r.downloadHistory[i].staticStartTime.UnixNano() >= after.UnixNano()
+		})
+	}
+
+	// Checking for cases when the same timestamp is submitted,
+	// so the user wants to delete a specific download
+	if beforeIndex == afterIndex {
+		// Checking for cases where the timestamp is not in the history
+		if afterIndex == len(r.downloadHistory) {
+			return nil
+		}
+		if !r.downloadHistory[afterIndex].staticStartTime.Equal(after) {
+			return nil
+		}
+	}
+
+	// This is for the case when the timestamp is not a download but is
+	// within the bounds of the array, we don't want the next download to be
+	// cleared since the user didn't submit that download's timestamp
+	if beforeIndex != len(r.downloadHistory) && !before.IsZero() && !r.downloadHistory[beforeIndex].staticStartTime.Equal(before) {
+		beforeIndex--
+	}
+
+	// Manually set indices if we didn't find any element.
+	if afterIndex == len(r.downloadHistory) && !before.IsZero() {
+		// Checking for before not equal to zero for case of only after
+		// is submitted but it is before all the timestamps
+		afterIndex = 0
+	}
+	if beforeIndex == len(r.downloadHistory) {
+		beforeIndex = len(r.downloadHistory) - 1
+	}
+
+	r.downloadHistory = append(r.downloadHistory[:afterIndex], r.downloadHistory[beforeIndex+1:]...)
+	return nil
```

Comment:
this function seems suspiciously complex. Given the number of edge cases, I'm strongly inclined to do away with the binary search optimization stuff and rewrite this in as clear a style as possible, which for me means an  O(n) loop:
```go
if len(r.downloadHistory) == 0 {
	return nil
}
if after.IsZero() {
	after = r.downloadHistory[0]
}
if before.IsZero() {
	before = r.downloadHistory[len(r.downloadHistory)-1]
}
withinTimespan := func(t time.Time) bool {
	return (t.After(after) || t.Equal(after)) && (t.Before(before) || t.Equal(before))
}
filtered := r.downloadHistory[:0]
for _, d := range r.downloadHistory {
	if !withinTimespan(d.staticStartTime) {
		filtered = append(filtered, d)
	}
}
r.downloadHistory = filtered
return nil
```
thoughts?

Reply:
I do agree that the function is a little complex for what it is trying to do.  If you think it's fine to go with an O(n) loop in order for the function to be more readable I can make those updates.

Reply:
ok, let's do it. We can always come back and optimize later if needed.